python/lsst/pipe/base/graphBuilder.py

```diff
@@ -530,12 +564,27 @@ def connectDataIds(self, registry, collections, userQuery, externalDataId):
         # associated with the input dataset types, but don't (yet) try to
         # obtain the dataset_ids for those inputs.
         _LOG.debug("Submitting data ID query and materializing results.")
-        with registry.queryDataIds(self.dimensions,
-                                   datasets=list(self.inputs),
-                                   collections=collections,
-                                   where=userQuery,
-                                   dataId=externalDataId,
-                                   ).materialize() as commonDataIds:
+        queryArgs = {'dimensions': self.dimensions, 'where': userQuery, 'dataId': externalDataId}
+        if datasetQueryConstraint == DatasetQueryConstraintVariant.ALL:
+            _LOG.debug("Constraining graph query using all datasets in pipeline.")
+            queryArgs['datasets'] = list(self.inputs)
+            queryArgs['collections'] = collections
+        elif datasetQueryConstraint == DatasetQueryConstraintVariant.OFF:
+            _LOG.debug("Not using dataset existance to constrain query.")
+        elif datasetQueryConstraint == DatasetQueryConstraintVariant.LIST:
```

Comment:
Feels weird to me that this variant instance satisfies an equality check against what (seems to be) its type. 

Feel free to defer fixing this until later, as I know the variant is going to change later anyway.  But it's something I think those fixes should address.